File: agentdock-plugin/src/main/kotlin/com/agentdock/terminal/CommandRenderer.kt
package com.agentdock.terminal

import com.agentdock.model.ProviderCommandContext

sealed class CommandRenderResult {
    data class Success(val command: String) : CommandRenderResult()
    data class MissingVariable(val variable: String) : CommandRenderResult()
}

class CommandRenderer(private val shellEscaper: ShellEscaper = ShellEscaper()) {
    private val variablePattern = Regex("\\{\\{([A-Za-z0-9_]+)(\\?)?}}")

    fun render(template: String, context: ProviderCommandContext): CommandRenderResult {
        val values = mapOf(
            "executable" to context.provider.executable,
            "providerSessionId" to context.session.providerSessionId,
            "sessionName" to context.session.name,
            "cwd" to context.session.cwd,
            "projectPath" to context.projectPath
        )

        var rendered = template
        for (match in variablePattern.findAll(template).toList()) {
            val variable = match.groupValues[1]
            val optional = match.groupValues[2] == "?"
            val rawValue = values[variable]
            if (rawValue.isNullOrBlank()) {
                if (optional) {
                    rendered = rendered.replace(match.value, "")
                    continue
                }
                return CommandRenderResult.MissingVariable(variable)
            }
            rendered = rendered.replace(match.value, shellEscaper.escape(rawValue, context.os))
        }

        return CommandRenderResult.Success(rendered.trim())
    }
}

File: AgentDock-VSCode/src/core/commandRenderer.ts
import type { ProviderCommandContext } from "./model.js";

export type CommandRenderResult =
  | { ok: true; command: string }
  | { ok: false; missingVariable: string };

const VARIABLE_PATTERN = /\{\{([A-Za-z0-9_]+)(\?)?}}/g;

export function renderCommand(template: string, context: ProviderCommandContext): CommandRenderResult {
  const values: Record<string, string | undefined> = {
    executable: context.provider.executable,
    providerSessionId: context.session.providerSessionId,
    sessionName: context.session.name,
    cwd: context.session.cwd,
    projectPath: context.projectPath
  };
  let missingVariable: string | undefined;
  const command = template.replace(VARIABLE_PATTERN, (placeholder, variable: string, optionalMark: string) => {
    const rawValue = values[variable];
    if (!rawValue?.trim()) {
      if (optionalMark === "?") {
        return "";
      }
      missingVariable ??= variable;
      return placeholder;
    }
    return escapeShellArgument(rawValue, context.platform);
  });

  if (missingVariable) {
    return { ok: false, missingVariable };
  }
  return { ok: true, command: command.trim() };
}

export function escapeShellArgument(value: string, platform: NodeJS.Platform): string {
  if (value.length === 0) {
    return platform === "win32" ? '""' : "''";
  }
  if (/^[A-Za-z0-9_./:=@%+-]+$/.test(value)) {
    return value;
  }
  if (platform === "win32") {
    return `"${value.replace(/(["^%])/g, "^$1")}"`;
  }
  return `'${value.replace(/'/g, `'"'"'`)}'`;
}
